### DISCUSSION TITLE
hid_device_info->interface_number zero on macOS 13.3

### DISCUSSION THREAD

**Discussion by dingsijie-nreal** [libusb/hidapi#527] [Q&A] [answered]
I've noticed comment here: https://github.com/libusb/hidapi/blob/master/mac/hid.c#L543
The interesting thing is: macOS Monterey(12.6.3) works perfectly with my usb gadget, I got expected interface numbers.
the output of `test.c` is shown below:
```
hidapi test/example tool. Compiled with hidapi version 0.14.0, runtime ve
Compile-time version matches runtime version of hidapi.

Device Found
  type: 3318 0424
  path: DevSrvsID:4298350859
  serial_number: A00016:33:33
  Manufacturer: xxxx
  Product:    xxxx
  Release:      1
  Interface:    3
  Usage (page): 0x0 (0x41)
  Bus type: 1

  Report Descriptor: (27 bytes)
0x05, 0x41, 0x09, 0x00, 0xa1, 0x01, 0x15, 0x00, 0x25, 0xff, 
0x19, 0x01, 0x29, 0x08, 0x95, 0x40, 0x75, 0x08, 0x81, 0x02, 
0x19, 0x01, 0x29, 0x08, 0x91, 0x02, 0xc0, 
Device Found
  type: 3318 0424
  path: DevSrvsID:4298350872
  serial_number: A00016:33:33
  Manufacturer: xxxx
  Product:      xxxxx
  Release:      1
  Interface:    5
  Usage (page): 0x0 (0x41)
  Bus type: 1

  Report Descriptor: (27 bytes)
0x05, 0x41, 0x09, 0x00, 0xa1, 0x01, 0x15, 0x00, 0x25, 0xff, 
0x19, 0x01, 0x29, 0x08, 0x95, 0x40, 0x75, 0x08, 0x81, 0x02, 
0x19, 0x01, 0x29, 0x08, 0x91, 0x02, 0xc0, 
Device Found
  type: 3318 0424
  path: DevSrvsID:4298350862
  serial_number: A00016:33:33
  Manufacturer: xxxx
  Product:      xxxxx
  Release:      1
  Interface:    4
  Usage (page): 0x0 (0x41)
  Bus type: 1

  Report Descriptor: (27 bytes)
0x05, 0x41, 0x09, 0x00, 0xa1, 0x01, 0x15, 0x00, 0x25, 0xff, 
0x19, 0x01, 0x29, 0x08, 0x95, 0x40, 0x75, 0x08, 0x81, 0x02, 
0x19, 0x01, 0x29, 0x08, 0x91, 0x02, 0xc0, 
```
But, after upgrading to macOS 13.3 the result changes to flowing:
```
hidapi test/example tool. Compiled with hidapi version 0.14.0, runtime version 0.14.0.
Compile-time version matches runtime version of hidapi.

Device Found
  type: 3318 0424
  path: DevSrvsID:4294984005
  serial_number: A00016:33:33
  Manufacturer: xxxxx
  Product:      xxxxx
  Release:      1
  Interface:    0
  Usage (page): 0x0 (0x41)
  Bus type: 1

  Report Descriptor: (27 bytes)
0x05, 0x41, 0x09, 0x00, 0xa1, 0x01, 0x15, 0x00, 0x25, 0xff, 
0x19, 0x01, 0x29, 0x08, 0x95, 0x40, 0x75, 0x08, 0x81, 0x02, 
0x19, 0x01, 0x29, 0x08, 0x91, 0x02, 0xc0, 
Device Found
  type: 3318 0424
  path: DevSrvsID:4294984010
  serial_number: A00016:33:33
  Manufacturer: xxxxx
  Product:      xxxxx
  Release:      1
  Interface:    0
  Usage (page): 0x0 (0x41)
  Bus type: 1

  Report Descriptor: (27 bytes)
0x05, 0x41, 0x09, 0x00, 0xa1, 0x01, 0x15, 0x00, 0x25, 0xff, 
0x19, 0x01, 0x29, 0x08, 0x95, 0x40, 0x75, 0x08, 0x81, 0x02, 
0x19, 0x01, 0x29, 0x08, 0x91, 0x02, 0xc0, 
Device Found
  type: 3318 0424
  path: DevSrvsID:4294984008
  serial_number: A00016:33:33
  Manufacturer: xxxxx
  Product:      xxxxx
  Release:      1
  Interface:    0
  Usage (page): 0x0 (0x41)
  Bus type: 1

  Report Descriptor: (27 bytes)
0x05, 0x41, 0x09, 0x00, 0xa1, 0x01, 0x15, 0x00, 0x25, 0xff, 
0x19, 0x01, 0x29, 0x08, 0x95, 0x40, 0x75, 0x08, 0x81, 0x02, 
0x19, 0x01, 0x29, 0x08, 0x91, 0x02, 0xc0, 
```
My mission it to discriminate 3 different components on my USB gadget, and the interface number was the key that I rely on.
But with 0 interfaces numbers, I don't know how to do my job.
Any one get any ideas? Thanks a lot.

**Comment by Youw** [comment]
Would you be able to modify HIDAPI locally and print all `dev` properties (key/values)?

**Reply by dingsijie-nreal** [reply]
Yes, I can. And I think all items in `hid_api_info` struct (https://github.com/libusb/hidapi/blob/master/hidapi/hidapi.h#L150)
already printed out.
What other info shall I try to print?

**Reply by Youw** [reply]
I meant printing all properties available via [IOHIDDeviceGetProperty](https://github.com/libusb/hidapi/blob/9f185ec51025695ab004085600ddaafb252e1e42/mac/hid.c#L307).
But I just briefly checked the API - it doesn't seem like there is a list of all available keys...

**Reply by dingsijie-nreal** [reply]
Well, I noticed this PR https://github.com/libusb/hidapi/pull/530
It was submitted by one of my coworkers.
The PR is a workaround we found.
For now I think https://github.com/libusb/hidapi/blob/9f185ec51025695ab004085600ddaafb252e1e42/mac/hid.c#L550 is the
only "proper" way to get the interface number. But it just doesn't work on the latest macOS.
We've posted a question at https://developer.apple.com/forums/thread/728001.
No one replys... Maybe you can also have a look at the post, and say something there.

**Reply by Youw** [reply]
> there is always a solution to find a corresponding USB/device interface having the Location ID - that one is guaranteed to match

that would be lesser workaround

**Comment by mcuee** [comment, marked as answer]
I have created an issue based on this discussion. I can reproduce the issue under macOS 13.3.1 on my Mac Mini M1. From my test, libusb works fine to find the interface number. Maybe we can take some reference from [libusb code](https://github.com/libusb/libusb/blob/master/libusb/os/darwin_usb.c#L1342-L1377).
* #531

**Reply by pengxianheng-nreal** [reply]
@mcuee I has tested by using libusb at first。 It really find non-zero interface number 。 but it  needs  to input 'sudo' to run an bin file or libusb_detach_kernel_driver will failed!

**Reply by mcuee** [reply]
Yes, that is correct. libusb is not a suitable library for HID device under macOS. In fact libusb project recommends using HIDAPI for HID device.

However, I am thinking that the following code snipets may be able to help to sort this issue out under macOS. I believe these lines of codes do not need sudo. sudo is only needed when you are carrrying out real USB transfers with the device, not for enumeration.
https://github.com/libusb/libusb/blob/master/libusb/os/darwin_usb.c#L1342-L1377

Same comments here.
https://github.com/libusb/hidapi/issues/531#issuecomment-1509600330

**Comment by Youw** [comment]
This is more a bug report than a discussion in a first place.
Closing in favor of #531.
